src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/StackOverflowClient.java

```diff
@@ -0,0 +1,143 @@
+    URI uri = new URI(url);
+    // Parse the URL to get the question id.
+    String[] segments = uri.getPath().split("/");
+    String questionId = segments[ID_INDEX];
+    if (!Pattern.matches("[0-9]+", questionId)) {
+      return null;
+    }
+    String searchUrl = String.format(SEARCH_URL_TEMPLATE, questionId);
+    try {
+      JSONObject res = getResponse(searchUrl);
+      String title =
+          res.getJSONArray(ITEM_PARAMETER).getJSONObject(0).get(TITLE_PARAMETER).toString();
+      card.setLink(uri.toString());
+      // Store the id of the question in order to get the code body of the answer.
+      card.setCode(questionId);
+      card.setTitle(title);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return card;
+  }
+
+  /* Get the most voted answer's id and store it in the card. */
+  private Card getAnswerId(Card card) {
+    String questionUrl = String.format(QUESTION_URL_TEMPLATE, card.getCode());
+    try {
+      JSONObject res = getResponse(questionUrl);
+      String answerId =
+          res.getJSONArray(ITEM_PARAMETER).getJSONObject(0).get(ANSWER_ID_PARAMETER).toString();
+      // Replace the question id by the answer id in order to retrieve the code body next.
+      card.setCode(answerId);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return card;
+  }
+
+  /* Get the content of the answer and store it in the card. */
+  private Card getAnswer(Card card) {
+    String answerUrl = String.format(ANSWER_URL_TEMPLATE, card.getCode());
+    try {
+      JSONObject res = getResponse(answerUrl);
+      String body =
+          res.getJSONArray(ITEM_PARAMETER).getJSONObject(0).get(BODY_PARAMETER).toString();
+      Document doc = Jsoup.parse(body);
+      // Combine all description in the answer body.
+      Elements descriptionHtml = doc.select("p");
+      String description = "";
+      for (Element e : descriptionHtml) {
+        description += e.outerHtml();
+      }
+      if (description.length() >= 200) {
+        description = description.substring(0, DESCRIPTION_LENGTH_PARAMETER);
+      }
+      // Combine all code in the answer body.
+      Elements codeHtml = doc.select(CODE_PARAMETER);
+      String code = "";
+      for (Element e : codeHtml) {
+        code += e.outerHtml();
+      }
+      card.setDescription(description);
+      card.setCode(code);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return card;
+  }
+
+  private JSONObject getResponse(String url) throws IOException {
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    CloseableHttpResponse response = httpClient.execute(new HttpGet(url));
+    if (response.getStatusLine().getStatusCode() != 200) {
+      return new JSONObject();
+    }
+    HttpEntity entity = response.getEntity();
+    if (entity == null) {
+      return new JSONObject();
+    }
+    BufferedReader reader = new BufferedReader(new InputStreamReader(entity.getContent()));
```

Comment:
I'm no expert in Java. But I think you need to be closing resources after you use them.